src/hierarchy_builder.rs

```diff
@@ -0,0 +1,206 @@
+            right: right,
+            left: left,
+            idx: index,
+        };
+        (s, &mut z[0])
+    }
+
+    pub fn get(&self, zindex: &ZoneIndex) -> &Zone {
+        let idx = zindex.index;
+        if idx < self.idx {
+            return &self.left[idx];
+        } else if idx == self.idx {
+            panic!("Cannot retrieve middle index");
+        } else {
+            return &self.right[idx - self.idx - 1];
+        }
+    }
+}
+
+// the goal is that f in [down(f as f32) as f64, up(f as f32) as f64]
+fn down(f: f32) -> f32 {
+    f - (f * ::std::f32::EPSILON).abs()
+}
+fn up(f: f32) -> f32 {
+    f + (f * ::std::f32::EPSILON).abs()
+}
+
+#[cfg(test)]
+mod test {
+    use geo::{LineString, MultiPolygon, Point, Polygon};
+    use zone::{Zone, ZoneIndex};
+    use hierarchy_builder::{build_hierarchy, ZonesTree};
+    use osmpbfreader::Tags;
+
+    fn zone_factory(idx: usize, ls: LineString<f64>, admin_level: u32) -> Zone {
+        let p = Polygon::new(ls, vec![]);
+        let mp = MultiPolygon(vec![p.clone()]);
+
+        Zone {
+            id: ZoneIndex { index: idx },
+            osm_id: "".into(),
+            admin_level: Some(admin_level),
+            admin_type: None,
+            name: "".into(),
+            center: None,
+            boundary: Some(mp),
+            parent: None,
+            tags: Tags::new(),
+            wikidata: None,
+            zip_codes: vec![],
+        }
+    }
+
+    #[test]
+    #[cfg_attr(rustfmt, rustfmt_skip)]
+    fn hierarchy_test() {
+        let l0 = LineString(vec![
+            Point::new(0., 0.),     //  +------+
+            Point::new(0., 10.),    //  |      |
+            Point::new(10., 10.),   //  |  z0  |
+            Point::new(10., 0.),    //  |      |
+            Point::new(0., 0.),     //  +------+
+        ]);
+        let z0 = zone_factory(0, l0, 2);
+
+        let l1 =LineString(vec![
+            Point::new(1., 1.),     //  +------+
+            Point::new(1., 9.),     //  |+----+|
+            Point::new(9., 9.),     //  || z1 ||
+            Point::new(9., 1.),     //  |+----+|
+            Point::new(1., 1.),     //  +------+
+        ]);
+        let z1 = zone_factory(1, l1, 5);
+
+        let l2 =LineString(vec![
+            Point::new(2., 2.),     //  +------+
+            Point::new(2., 8.),     //  |      |
+            Point::new(8., 8.),     //  |  []<---- z2
+            Point::new(8., 2.),     //  |      |
+            Point::new(2., 2.),     //  +------+
+        ]);
+        let z2 = zone_factory(2, l2, 6);
+
+        let l3 =LineString(vec![
+            Point::new(0., 0.),     //  +------+
+            Point::new(0., 5.),     //  |      |
+            Point::new(10., 5.),    //  +------+
+            Point::new(10., 0.),    //  |  z3  |
+            Point::new(0., 0.),     //  +------+
+        ]);
+        let z3 = zone_factory(2, l3, 3);
+
+        let mut zones = vec![z0,z1,z2,z3];
+        let mut ztree = ZonesTree::new();
```

Comment:
I think it would be simplier to implement the trait `FromIterator`

(like [here](https://github.com/CanalTP/mimirsbrunn/blob/master/src/admin_geofinder.rs#L149-L159) )